nexus/reconfigurator/planning/src/blueprint_builder.rs

```diff
@@ -405,6 +438,89 @@ impl<'a> BlueprintBuilder<'a> {
+    /// - Ensure that any disks / zpools that exist in the database
+    ///   are propagated into the blueprint.
+    /// - Ensure that any disks that are expunged from the database are
+    ///   removed from the blueprint.
+    pub fn sled_ensure_disks(
+        &mut self,
+        sled_id: TypedUuid<SledKind>,
+        resources: &SledResources,
+    ) -> Result<Ensure, Error> {
+        let (mut additions, removals) = {
+            // These are the disks known to our (last?) blueprint
+            let blueprint_disks: BTreeMap<_, _> = self
+                .disks
+                .current_sled_disks(sled_id)
+                .map(|disk| {
+                    (
+                        TypedUuid::<PhysicalDiskKind>::from_untyped_uuid(
+                            disk.id,
+                        ),
+                        disk,
+                    )
+                })
+                .collect();
+
+            // These are the in-service disks as we observed them in the database,
+            // during the planning phase
+            let database_disks: BTreeMap<_, _> = resources
+                .all_disks(DiskFilter::InService)
+                .map(|(zpool, disk)| (disk.disk_id, (zpool, disk)))
+                .collect();
+
+            // Add any disks that appear in the database, but not the blueprint
+            let additions = database_disks
+                .iter()
+                .filter_map(|(disk_id, (zpool, disk))| {
+                    if !blueprint_disks.contains_key(disk_id) {
+                        Some(BlueprintPhysicalDiskConfig {
+                            identity: disk.disk_identity.clone(),
+                            id: disk_id.into_untyped_uuid(),
+                            pool_id: zpool.into_untyped_uuid(),
+                        })
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<BlueprintPhysicalDiskConfig>>();
+
+            // Remove any disks that appear in the blueprint, but not the database
+            let removals: HashSet<TypedUuid<PhysicalDiskKind>> =
+                blueprint_disks
+                    .keys()
+                    .filter_map(|disk_id| {
+                        if !database_disks.contains_key(disk_id) {
+                            Some(*disk_id)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+
+            (additions, removals)
+        };
+
+        if additions.is_empty() && removals.is_empty() {
+            return Ok(Ensure::NotNeeded);
+        }
+
+        let disks = &mut self.disks.change_sled_disks(sled_id).disks;
+
+        disks.append(&mut additions);
+        disks.retain(|config| {
+            !removals.contains(
+                &TypedUuid::<PhysicalDiskKind>::from_untyped_uuid(config.id),
+            )
+        });
+
+        Ok(Ensure::Added)
```

Comment:
Nit: maybe not worth changing right now since this is used elsewhere, but `Ensure::Modified` is perhaps a better name.